backend/app/checkpoint.py

```diff
@@ -1,147 +1,112 @@
-import pickle
-from datetime import datetime
-from typing import AsyncIterator, Optional
+import os
+from typing import Any, AsyncIterator, Optional, Sequence
 
-from langchain_core.messages import BaseMessage
-from langchain_core.runnables import ConfigurableFieldSpec, RunnableConfig
-from langgraph.checkpoint.base import BaseCheckpointSaver
+import structlog
 from langgraph.checkpoint.base import (
+    ChannelVersions,
     Checkpoint,
-    CheckpointAt,
-    CheckpointThreadTs,
+    CheckpointMetadata,
     CheckpointTuple,
-    SerializerProtocol,
+    RunnableConfig,
 )
+from langgraph.checkpoint.postgres.aio import AsyncPostgresSaver
+from langgraph.checkpoint.postgres.base import BasePostgresSaver
+from langgraph.checkpoint.serde.base import SerializerProtocol
+from psycopg import AsyncPipeline
+from psycopg_pool import AsyncConnectionPool
 
-from app.lifespan import get_pg_pool
+logger = structlog.get_logger(__name__)
 
 
-def loads(value: bytes) -> Checkpoint:
-    loaded: Checkpoint = pickle.loads(value)
-    for key, value in loaded["channel_values"].items():
-        if isinstance(value, list) and all(isinstance(v, BaseMessage) for v in value):
-            loaded["channel_values"][key] = [v.__class__(**v.__dict__) for v in value]
-    return loaded
+class AsyncPostgresCheckpoint(BasePostgresSaver):
```

Comment:
Have you considered initializing on app start up and calling `.setup()`  to set up the migration, and then avoiding doing the wrapping of the checkpointer?

It'll help keep the checkpoints in sync and remove some extra code here

Reply:
Hey Eugene,

**Checkpoint:**

I’ve run into a few challenges while implementing langgraph’s checkpoint:
1.	Global Checkpointer Initialization:
I defined the checkpointer in the lifespan and declare it as a global in `agent.py`. However, I encountered the error “Checkpointer not initialized” because the global instance wasn’t properly initialized before being accessed.
This issue occurs because the checkpointer depends on the application startup completing before it can be used.

2.	Singleton Pattern Issues:
I tried using a singleton pattern for `AsyncPostgresSaver` to manage the global instance, initializing it during the lifespan. However, the initialization of `AsyncPostgresSaver` requires an async event loop, which isn’t always available—such as during testing—resulting in the error: “no running event loop.”

3. Current Implementation:
I implemented a solution inspired by the current approach in OpenGPTs, adapted to use the new checkpointer in LangGraph:
- Singleton with Lazy Initialization: Created a `BasePostgresSaver` class with a singleton pattern that assigns the instance before initialization.
- Async Setup Method: Moved the connection pool creation to an async `setup()` method, ensuring it initializes during the lifespan of the application when an asynchronous loop is available.
 
I’m open to trying a different approach if you have any suggestions or recommendations!

**Migration:**

Finally, I decided to run the migrations before using the app to stay consistent with OpenGPTs and ensure the queries are ready. I am using the same .sql.

What do you think?